src/core/auth/storage/qgsauthconfigurationstoragedb.cpp

```diff
@@ -0,0 +1,2026 @@
+    authdb = QSqlDatabase::addDatabase( mDriver, connectionName );
+
+
+    // Check that the driver is available
+    if ( !QSqlDatabase::isDriverAvailable( mDriver ) )
+    {
+      const_cast< QgsAuthConfigurationStorageDb * >( this )->setError( tr( "Driver '%1' is not available" ).arg( mDriver ) );
+      return authdb;
+    }
+
+    authdb.setDatabaseName( mDatabase );
+    authdb.setHostName( mHost );
+    authdb.setPort( mPort.toInt() );
+    authdb.setUserName( mUser );
+    authdb.setPassword( mPassword );
+    authdb.setConnectOptions( mConnectOptions );
+
+    // Return if not valid
+    if ( !authdb.isValid() )
+    {
+      const_cast< QgsAuthConfigurationStorageDb * >( this )->setError( tr( "Auth db connection is not valid" ) );
+      return authdb;
+    }
+
+    // for background threads, remove database when current thread finishes
+    if ( QCoreApplication::instance() && QThread::currentThread() != QCoreApplication::instance()->thread() )
+    {
+      QgsDebugMsgLevel( QStringLiteral( "Scheduled auth db remove on thread close" ), 2 );
+
+      // IMPORTANT - we use a direct connection here, because the database removal must happen immediately
+      // when the thread finishes, and we cannot let this get queued on the main thread's event loop (where
+      // QgsAuthManager lives).
+      // Otherwise, the QSqlDatabase's private data's thread gets reset immediately the QThread::finished,
+      // and a subsequent call to QSqlDatabase::database with the same thread address (yep it happens, actually a lot)
+      // triggers a condition in QSqlDatabase which detects the nullptr private thread data and returns an invalid database instead.
+      // QSqlDatabase::removeDatabase is thread safe, so this is ok to do.
+      // Right about now is a good time to re-evaluate your selected career ;)
+      // I've done that and I decided to become a musician. I'll probably be a better musician than a software developer.
+      QMetaObject::Connection connection = connect( QThread::currentThread(), &QThread::finished, QThread::currentThread(), [connectionName, this ]
+      {
+        QMutexLocker locker( &mMutex );
+        QSqlDatabase::removeDatabase( connectionName );
+        mConnectedThreads.remove( QThread::currentThread() );
+      }, Qt::DirectConnection );
+
+      mConnectedThreads.insert( QThread::currentThread(), connection );
+    }
+  }
+  else
+  {
+    QgsDebugMsgLevel( QStringLiteral( "Reusing existing connection" ), 2 );
+    authdb = QSqlDatabase::database( connectionName, false );
+  }
+
+  locker.unlock();
+
+  if ( !authdb.isOpen() )
+  {
+    if ( !authdb.open() )
+    {
+      const_cast< QgsAuthConfigurationStorageDb * >( this )->setError( tr( "Opening of authentication db FAILED : %1" ).arg( authdb.lastError().text() ) );
+    }
+  }
+
+  return authdb;
+}
+
+bool QgsAuthConfigurationStorageDb::authDbOpen() const
+{
+  QMutexLocker locker( &mMutex );
```

Comment:
Are all these implementations 1:1 copies of the current code? Or are there modifications I should look over specifically?

Reply:
They are mostly 1:1 copies but I did not just cut & paste. I'd say that you can defer the review of the CRUD methods after finish writing the tests, I will probably intercept most issues while testing.